### PR TITLE
VIP-10: Storage node registration (on-chain)

### PR DESCRIPTION
**Pull request by tahpot**

### CODE REVIEW

File: VIPs/vip-10.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
```
Comment (tahpot) [inline VIPs/vip-10.md:28]: @ITStar10 Can you propose the appropriate data structures

File: VIPs/vip-10.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
```
Comment (tahpot) [inline VIPs/vip-10.md:50]: @ITStar10 Can you propose appropriate state variables

File: VIPs/vip-10.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
+6. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+An `establishmentDate` will be saved that matches the current unix timestamp (`block.timestamp`).
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
```
Comment (ITStar10) [inline VIPs/vip-10.md:67]: I think, `unregisteredDatetime` is not necessary because we can get time in the blockchain itself.

File: VIPs/vip-10.md (hunk 2 above)
Comment (ITStar10) [reply VIPs/vip-10.md:50]: - `countryCode` and `regionCode` are duplicated between StorageNode & Datacenter. Or are tehy different values?

Comment (ITStar10) [reply VIPs/vip-10.md:50]: This is the draft of state variables:
```
struct StorageNode {
  address didAddress;
  string endpointUri;
  uint datacenterId;

  uint unregisterTime;
}

struct Datacenter {
  string countryCode;
  string regionCode;
  uint lat;
  uint long;
}

/** mapping of storageNode ID => Storage*/
map(uint => StorageNode) private _storageNodeMap;
/** mapping of did address => storageNode ID*/
map(address => uint) private _addressNodeMap;
/** mapping of endpoint uri => storageNode ID*/
map(string => uint) private _endpointNodeMap;
/** mapping of countryCode => StorgeNode ID list*/
map(string => EnumerableSet.UintSet) private _countryNodes;
/** mapping of regionCode => StorageNode ID list*/
map(string => EnumerableSet.UintSet) private _regionNodes;

/** mapping of datacenterId => DataCenter */
map(uint => Datacenter) dataCenterMap
/** mapping of datacenterID => StorageNode ID list*/
map(uint => EnumerableSet.UintSet) private _dataCenterToNodeIdsmap;


CountersUpgradeable.Counter private _nodeIdCounter;
CountersUpgradeable.Counter private _dataCenterIdCounter;
```

**Comment by ITStar10** [comment]
@tahpot How can we name this contract?<br/>
Here are my suggestions:
- contract name: `StorageNodeContract` - package name:  `@verida/storage-node-contract`
- contract name: `VeridaNodeContract` - package name:  `@verida/node-contract`

**Comment by tahpot** [comment]
@ITStar10 

I would like:

- Contract name: `StorageNodeRegistry`
- Package name: `@verida/storage-node-client`

File: VIPs/vip-10.md (hunk 2 above)
Comment (tahpot) [reply VIPs/vip-10.md:50]: >countryCode and regionCode are duplicated between StorageNode & Datacenter. Or are tehy different values?

Correct, duplicated.

If you have them in your mapping it's okay to only store them in `Datacenter`.

**Comment by ITStar10** [comment]
@tahpot `@verida/storage-node-client` is for verida-js?

**Comment by ITStar10** [comment]
`countryCode` and `regionCode` in the StorageNode & Datacenter are the same?

**Comment by tahpot** [comment]
> `countryCode` and `regionCode` in the StorageNode & Datacenter are the same?

No, different.

File: VIPs/vip-10.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
```
Comment (ITStar10) [inline VIPs/vip-10.md:56]: - A `didAddress` can be used only once for `addNode()` function?
- `endpointUri` is also unique in the contract? There can't be any same `endpointUri` with different `countryCode` or `regionCode`?

Comment (tahpot) [reply VIPs/vip-10.md:56]: 1. Yes
2. Yes

File: VIPs/vip-10.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+## removeNode(didAddress: string, unregisterDatetime: uint, requestSignature: string)
+
+Unregister a storage node from the network at the specified date:
+
+1. `didAddress`: The DID that is to be removed from the network
+2. `unregisterDatetime`: The unix timestamp of when the storage node should no longer be available for selection. Must be at least 28 days in the future.
+3. `requestSignature`: The request parameters signed by the `didAddress` private key. Will be verified by `VDA-Verification-Base` library.
+
+Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.
```
Comment (aurelticot) [inline VIPs/vip-10.md:79]: What's the motivation behind the 28-day notice (unregisterDatetime) if the `removeNode` is "just" disabling the discoverability?

File: VIPs/vip-10.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+Add a data center to the network. `id` will be autoincremented.
+
+Returns `int` the unique ID of the created data center
+
+## removeDatacenter(id: string) ownerOnly
+
+Remove a data center by `id`. Will only remove the data center if there are no storage nodes using that datacenter.
+
+## getDatacenters(ids: string[]): Datacenter[]
```
Comment (aurelticot) [inline VIPs/vip-10.md:105]: For Storage Nodes to register themselves, they will have to pass the Id of their datacenter.
This `getDatacenters(ids)` taking a list of ids as parameter, doesn't make it easy for them to find their exact datacenter?
Maybe a `getAllDatacenters()` or a `getDatacentersByCountry(countryCode)`

File: VIPs/vip-10.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
+
+This method registers a new endpoint on the network:
+
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
```
Comment (aurelticot) [inline VIPs/vip-10.md]: > datacenter previously created with `addDatacenter()`

Feels like `addDatacenter` must be called by the Storage Node prior to call `addNode`, but `addDatacenter` is `ownerOnly`.

Maybe reword as "must match an existing datacenter, available via `getDatacenters`"

File: VIPs/vip-10.md (hunk 3 above)
Comment (aurelticot) [inline VIPs/vip-10.md]: If a Storage Node wants to temporarily disable its discoverability, and so call `removeNode`, could it be able to register itself later with the same `didAddress`? Or does it have to use a different `didAddress`?

What's the impact (for the Storage Node and associated data, ...) if a different `didAddress` is needed?

Particularly looking at [@ITStar10 comment](https://github.com/verida/VIPs/pull/4/files#r1173420202) with the `StorageNode struct` defining a `unregisterTime property`

File: VIPs/vip-10.md (hunk 7 above)
Comment (tahpot) [reply VIPs/vip-10.md]: No, this is intentional.

At this stage, only we should be able to add data centres to avoid duplications etc.

In the future there could be a more complex decentralized voting process to add that type of info.

File: VIPs/vip-10.md (hunk 3 above)
Comment (tahpot) [reply VIPs/vip-10.md]: >If a Storage Node wants to temporarily disable its discoverability, and so call removeNode, could it be able to register itself later with the same didAddress? Or does it have to use a different didAddress?

Provided there is no active node registered with that `didAddress`, it should be able to register again with the same address.

File: VIPs/vip-10.md (hunk 5 above)
Comment (tahpot) [reply VIPs/vip-10.md:79]: Good question. See latest commit where I've added more detail.

File: VIPs/vip-10.md (hunk 6 above)
Comment (tahpot) [reply VIPs/vip-10.md:105]: Good point.

Perhaps we add:

- `getDatacentersByCountry(countryCode: string): DataCenter[]`
- `getDatacentersByRegion(regionCode: string): DataCenter[]`

Thoughts @itstar?

File: VIPs/vip-10.md (hunk 2 above)
Comment (tahpot) [reply VIPs/vip-10.md:50]: @ITStar10 Can you push a commit to add your proposed state variables?

File: VIPs/vip-10.md (hunk 7 above)
Comment (aurelticot) [reply VIPs/vip-10.md]: Yep, I get the `ownerOnly` for creating a datacenter.
I was pointing the "previously created" may suggest they (the storage nodes) have to create the datacenter, but they should not (they can't), they must use an existing one.

File: VIPs/vip-10.md (hunk 3 above)
Comment (aurelticot) [reply VIPs/vip-10.md]: Ok, good.

Then, I'm wondering about the `unregisterTime` in @ITStar10 comment. It made me think the record won't be deleted but "flagged" as unregistered/inactive

File: VIPs/vip-10.md (hunk 5 above)
Comment (aurelticot) [reply VIPs/vip-10.md:79]: But, as the note suggests the discoverability and the operation (being active and managing data) of the node are **not related**, the existing users of the node are not affected by the discoverability or not.

Having a notice period if the node is shutting down, yes, because users have to move their data.

But for chaining the discoverability, I don't see the point.
I'd argue, Storage Nodes may want to be non-discoverable instantly in some cases (technical issues, legal issues, ...) so they don't have more users coming in.

File: VIPs/vip-10.md (hunk 6 above)
Comment (ITStar10) [reply VIPs/vip-10.md:105]: Ok, will add these functions.

File: VIPs/vip-10.md (hunk 3 above)
Comment (ITStar10) [reply VIPs/vip-10.md]: That was a draft of `StorageNode` struct. <br/>
I am going to remove the `unregisteredTime` in the struct and create an individual variable.

Comment (aurelticot) [reply VIPs/vip-10.md]: > That was a draft of `StorageNode` struct. I am going to remove the `unregisteredTime` in the struct and create an individual variable.

Ok. So that means the did is removed from the map(s) when calling `removeNode` so it can be added again later with `addNode`, right?

I mean, we don't have to go in the technical details in the spec, but the functional requirement is that a Storage Node can unregister the node and later register it again with the same DID.

File: VIPs/vip-10.md (hunk 2 above)
Comment (ITStar10) [reply VIPs/vip-10.md:50]: It might be better to push the State Variables after contract development done.

**Comment by ITStar10** [comment]
## Remove datacenter ID
We should remove datacenter only when there are no storage nodes depending on. <br/>
What happened in following case?<br/>
We created a datacenter. And we added a storage node using this datacenter. After a while, we removed storage node.<br/>
**Can we remove the datacenter in case**?

File: VIPs/vip-10.md (hunk 2 above)
Comment (tahpot) [reply VIPs/vip-10.md:50]: @ITStar10 This is a draft, so put what you have in here for now and we will update at the end.

**Comment by tahpot** [comment]
> @tahpot `@verida/storage-node-client` is for verida-js?

Yes

File: VIPs/vip-10.md (hunk 6 above)
Comment (tahpot) [reply VIPs/vip-10.md:105]: @ITStar10 Please add them to this draft spec document as well

**Comment by tahpot** [comment]
> ## Remove datacenter ID
> We should remove datacenter only when there are no storage nodes depending on. What happened in following case? We created a datacenter. And we added a storage node using this datacenter. After a while, we removed storage node. **Can we remove the datacenter in case**?

No, data centres shouldn't be automatically deleted.

File: VIPs/vip-10.md (hunk 5 above)
Comment (tahpot) [reply VIPs/vip-10.md:79]: I'm not following.

Can you suggest a change to illustrate what you're thinking?

File: VIPs/vip-10.md (hunk 7 above)
Comment (tahpot) [reply VIPs/vip-10.md]: Got it. Adjusted wording.

File: VIPs/vip-10.md (hunk 5 above)
Comment (aurelticot) [reply VIPs/vip-10.md:79]: @tahpot, this passage:

> 2. `unregisterDatetime`: [...] will be removed from the network. [...] ensure users have sufficient time to migrate away [...]. [...] expect the node to become unavailable and their data to be deleted from `unregisterDatetime` onwards.

and this passage:

> Note: This does not mean the node is no longer active and managing data for users on the network. It just means it is no longer discoverable for new connections.

are contradictory.

I personally lean towards the Note:

- Unregistering a node **only** disables its discoverability for new users. No notice period because it doesn't affects its users.
- Shutting down a node is a different process. It requires a notice period because users must move away.

Comment (tahpot) [reply VIPs/vip-10.md:79]: Hmm... I see what you mean, although I didn't intend it to come across as contradictory.

See the updated wording.

We don't want to support the concept of removing discoverability, but the node continues to operate for users. That will break the economic model when users are required to pay for storage via tokens. This contract is the source of truth of who is paying who.
